src/python/pants/engine/fs.py

```diff
@@ -170,13 +171,32 @@ class DirectoryWithPrefixToAdd:
 
 class DirectoryToMaterialize(datatype([('path', str), ('directory_digest', Digest)])):
   """A request to materialize the contents of a directory digest at the provided path."""
-  pass
+
+
+DirectoriesToMaterialize = Collection.of(DirectoryToMaterialize)
+
+
+class MaterializeDirectoryResult(datatype([('output_paths', string_list)])):
+  """Result of materializing a directory, contains the full output paths."""
+
+
+MaterializeDirectoriesResult = Collection.of(MaterializeDirectoryResult)
 
 
 class UrlToFetch(datatype([('url', str), ('digest', Digest)])):
   pass
 
 
+class Workspace:
```

Comment:
I think this should be a `dataclass`, unless there's a strong reason to have it be a normal class. Dataclass gives us a bunch of nice things for free like a good `__repr__`. When frozen, it also implements the requirements we need for the V2 engine like ~immutability.

```
from dataclasses import dataclass
from typing import Any

@dataclass(frozen=True)
class Workspace:
  _scheduler: Any
```
++ if you can get the type hint to be more precise than `Any`, but not blocking.